src/stdio.rs

```diff
@@ -0,0 +1,93 @@
+//! Enables direct write access to I/O streams in Python's `sys` module.
+
+//! In some cases printing to Rust's `std::io::stdout` or `std::io::stderr` will not appear
+//! in the Python interpreter, e.g. in Jupyter notebooks. This module provides a way to write
+//! directly to Python's I/O streams from Rust in such cases.
+
+//! ```rust
+//! use std::io::Write;
+//!
+//! let mut stdout = pyo3::stdio::stdout();
+//!   
+//! // This may not appear in Jupyter notebooks...
+//! println!("Hello, world!");
+//!
+//! // ...but this will.
+//! writeln!(stdout, "Hello, world!").unwrap();
+//! ```
+
+use crate::ffi::{PySys_WriteStderr, PySys_WriteStdout};
+use crate::Python;
+use std::io::{LineWriter, Write};
+use std::os::raw::c_char;
+
+macro_rules! make_python_stdio {
+    ($rawtypename:ident, $typename:ident, $pyfunc:ident, $stdio:ident) => {
+        struct $rawtypename {
+            cbuffer: Vec<u8>,
+        }
+        impl $rawtypename {
+            fn new() -> Self {
+                Self {
+                    cbuffer: Vec::new(),
+                }
+            }
+        }
+        impl Write for $rawtypename {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                //clear internal buffer and then overwrite with the
+                //new buffer and a null terminator
+                self.cbuffer.clear();
+                self.cbuffer.extend_from_slice(buf);
+                self.cbuffer.push(0);
+                Python::with_gil(|_py| unsafe {
+                    $pyfunc(self.cbuffer.as_ptr() as *const c_char);
+                });
+                Ok(buf.len())
+            }
+            fn flush(&mut self) -> std::io::Result<()> {
+                // call the python flush() on sys.$pymodname
+                Python::with_gil(|py| -> std::io::Result<()> {
+                    py.run_bound(
+                        std::concat!("import sys; sys.", stringify!($stdio), ".flush()"),
```

Comment:
Is flushing really necessary or rather what is the buffering behaviour of the `Write/Format` functions?

And if it is necessary, can we avoid repeated calling into the import machinery here, e.g. acquire the `Py<PyAny>` for `sys.stdout/err` once during construction and stash it in `Stdout/err`.

Reply:
Apparently, the provided functions really do not flush:

https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Python/sysmodule.c#L3895

So I guess, grabbing and stashing a reference to `sys.stdout/err` would be the way to go.

Reply:
I have rewritten things to hold `sys.stdout` or `sys.stderr` as appropriate, and then just call `flush` directly on the stored object.  

`flush` is part of the required implementation for the `Write` trait, thoug I suppose it could have been written as a noop.   There is no buffering if `Write` is just implemented directly for the output stream, which is why I have the `LineWriter` wrapper around the raw streams.   The `LineWriter` (as far as I understand) buffers inputs until a newline appears, and then calls `write` on its inner object.   The internal Rust `std::io::Stdout` type does[ something similar](https://github.com/rust-lang/rust/blob/a09d91b04a32def152afee94ed47933c03a84d0c/library/std/src/io/stdio.rs#L545).